umbraco/Umbraco-CMS
# Delivery API: Query controller

## Details
- Adding a new Examine index - ContentAPIIndex (also available through the Examine Management dashboard):
  - With a new FieldDefinitionCollection - not all the fields are there;
  - Currently **without** a Validator - there is a FIXME;
  - And a few other general TODOs - currently the index is not repopulated on publishing automatically.
- Creating a querying endpoint - `/content`:
  - There are 3 handlers to handle the different supported query parameters - they are used to collect all Selectors (those that can handle a _fetch_ query), Filters and Sorts;
    - The Handlers help construct our own query specification models that can be used when building _(translating to)_ Examine queries or any other indexing/searching engine;
    - There is a new querying service that will return a collection of ids of the items that matched the given search criteria based on which we can create our API content response per usual. 
  - You can either get everything from the index (_all nodes left-aligned in the Content tree_) without providing any query params, or you can add:
    - `?fetch={children / descendants / ancestors}:{id / path}`, 
    - `?filter=contentType:{alias}` and an additional `name={nodeName}` filter for testing combination of several filters,
    - `?sort={name / level / sortOrder / path}:{asc / desc}` - also for testing combination of several sorts.
- There are a few other small fixes for the other API controllers.

## Testing
- _In general:_ Set up some content items with different content types and different properties and make sure you can still apply the current params or headers that worked for the single item-based endpoints - like `?expand` param or `start-item` header;
- Test that `/content` just returns all content nodes left-aligned in the Content tree;
- Test different combinations of query params:
  - Add at least one of each - <code>/content?<b>fetch</b>=xxx&<b>filter</b>=yyy&<b>sort</b>=zzz</code>;
  - Add 1+ filters and/or sorts - <code>/content?<b>fetch</b>=xxx&<b>filter</b>=yyy&<b>filter</b>=vvv&<b>sort</b>=zzz&<b>sort</b>=www</code>;
  - Don't add `fetch` param and make sure you are filtering and sorting all items from the index;
    - Test the `fetch` param with both id and path;
  - Without `sort` param - the returned items will be returned in a default sort combination of `path` and `sortOrder`;
  - Test the current selectors/filters and sort implementations - if you get the expected results;
    - The two filters support negation at the moment - so you can negate the value like `filter=contentType:!article`  as “_anything but articles_”
- Test with query params that don't exist - we should get an empty collection;
  - Also when the value after `=` is invalid or there are no matches, we will get an empty collection.
